src/java/org/apache/cassandra/index/sai/disk/vector/CompactionGraph.java

```diff
@@ -0,0 +1,349 @@
+
+public class CompactionGraph implements Closeable, Accountable
+{
+    private static final Logger logger = LoggerFactory.getLogger(CompactionGraph.class);
+    private static final VectorTypeSupport vts = VectorizationProvider.getInstance().getVectorTypeSupport();
+
+    private final GraphIndexBuilder builder;
+    private final VectorType.VectorSerializer serializer;
+    private final VectorSimilarityFunction similarityFunction;
+    private final ChronicleMap<VectorFloat<?>, CompactionVectorPostings> postingsMap;
+    private final InlineVectorValues inlineVectors;
+    private final PQVectors pqVectors;
+    private final ArrayList<ByteSequence<?>> pqVectorsList;
+    private final IndexDescriptor descriptor;
+    private final IndexContext context;
+    private final boolean unitVectors;
+    private final int entriesAllocated;
+    private final File postingsFile;
+    private boolean postingsOneToOne;
+    private int nextOrdinal = 0;
+    private final ProductQuantization compressor;
+    private final OnDiskGraphIndexWriter writer;
+    private final long termsOffset;
+
+    public CompactionGraph(IndexDescriptor descriptor, IndexContext context, ProductQuantization compressor, boolean unitVectors, long keyCount) throws IOException
+    {
+        this.descriptor = descriptor;
+        this.context = context;
+        this.unitVectors = unitVectors;
+        var indexConfig = context.getIndexWriterConfig();
+        var termComparator = context.getValidator();
+        int dimension = ((VectorType<?>) termComparator).dimension;
+
+        // We need to tell Chronicle Map (CM) how many entries to expect.  it's critical not to undercount,
+        // or CM will crash.  However, we don't want to just pass in a max entries count of 2B, since it eagerly
+        // allocated segments for that many entries, which takes about 25s.
+        //
+        // If our estimate turns out to be too small, it's not the end of the world, we'll flush this segment
+        // and start another to avoid crashing CM.  But we'd rather not do this because the whole goal of
+        // CompactionGraph is to write one segment only.
+        var dd = descriptor.descriptor;
+        var rowsPerKey = Keyspace.open(dd.ksname).getColumnFamilyStore(dd.cfname).getMeanRowsPerPartition();
+        long estimatedRows = (long) (1.1 * keyCount * rowsPerKey); // 10% fudge factor
+        int maxRowsInGraph = Integer.MAX_VALUE - 100_000; // leave room for a few more async additions until we flush
+        entriesAllocated = estimatedRows > maxRowsInGraph ? maxRowsInGraph : (int) estimatedRows;
+
+        serializer = (VectorType.VectorSerializer) termComparator.getSerializer();
+        similarityFunction = indexConfig.getSimilarityFunction();
+        postingsOneToOne = true;
+        this.compressor = compressor;
+
+        Component tmpComponent = new Component(Component.Type.CUSTOM, "chronicle-tmp");
+        postingsFile = descriptor.descriptor.fileFor(tmpComponent);
```

Comment:
My only concern is what will happen if we exit without being able to call `close` on this class. Will we naturally clean it up on restart? I am unfamiliar with the lifecycle of these files, so my concern might be unfounded.

Reply:
i don't know, but if it's a problem it already exists for all the other components, so i'm inclined to say we don't have to solve it here

Reply:
... then again the other components were traditionally written all at once at flush/writeData time so the window for sadness was much smaller

Reply:
```java
        // We need to tell Chronicle Map (CM) how many entries to expect.  it's critical not to undercount,
        // or CM will crash.  However, we don't want to just pass in a max entries count of 2B, since it eagerly
        // allocated segments for that many entries, which takes about 25s.
```

This is the part that has me thinking about crashes. Is it enough to delete the temp file on start up?

Reply:
okay, i found the code that handles cleanup-unfinished-compaction and gave the file here an extension that will be recognized.  tested manually, it seems to work.
```
INFO  [main] 2024-05-01 10:26:48,903 ColumnFamilyStore.java:899 - Removing unfinished temporary file /home/jonathan/Projects/cassandra/data/data/pw/embeddings_table_new-49774b10fb3411eeaa92cd6e1ce4c599/chronicle.tmp
```

Reply:
Cool, TIL